src/dma/mod.rs

```diff
@@ -0,0 +1,798 @@
+                }
+                // Check if an overrun occurred, the buffer address won't be
+                // updated in that case
+                if self.stream.get_memory_address() != new_buf_ptr as u32 {
+                    self.stream.clear_transfer_complete_interrupt();
+                    return Err(DMAError::Overrun(new_buf));
+                }
+
+                // "Subsequent reads and writes cannot be moved ahead of
+                // preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                let old_buf = self.buf.replace(new_buf);
+
+                // We always have a buffer, so unwrap can't fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer));
+            } else {
+                unsafe {
+                    self.stream
+                        .set_memory_double_buffer_address(new_buf_ptr as u32);
+                }
+                // Check if an overrun occurred, the buffer address won't be
+                // updated in that case
+                if self.stream.get_memory_double_buffer_address()
+                    != new_buf_ptr as u32
+                {
+                    self.stream.clear_transfer_complete_interrupt();
+                    return Err(DMAError::Overrun(new_buf));
+                }
+
+                // "Subsequent reads and writes cannot be moved ahead of
+                // preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                let old_buf = self.double_buf.replace(new_buf);
+
+                // double buffering, unwrap can never fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::DoubleBuffer));
+            }
+        }
+        self.stream.disable();
+        self.stream.clear_transfer_complete_interrupt();
+
+        // "No re-ordering of reads and writes across this point is allowed"
+        compiler_fence(Ordering::SeqCst);
+
+        // NOTE(unsafe) We now own this buffer and we won't call any &mut
+        // methods on it until the end of the DMA transfer
+        let buf_len = unsafe {
+            let (buf_ptr, buf_len) = new_buf.write_buffer();
+
+            self.stream.set_memory_address(buf_ptr as u32);
+            buf_len
+        };
+        self.stream.set_number_of_transfers(buf_len as u16);
+        let old_buf = self.buf.replace(new_buf);
+
+        // Ensure that all transfers to normal memory complete before
+        // subsequent memory transfers.
+        //
+        // The new memory buffer is almost certainly in normal memory, so we
+        // ensure that all transfers there complete before proceeding to enable
+        // the stream
+        cortex_m::asm::dmb();
+
+        // "Preceding reads and writes cannot be moved past subsequent writes"
+        compiler_fence(Ordering::Release);
+
+        unsafe {
+            self.stream.enable();
+        }
+
+        Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer))
```

Comment:
In cases where the user is stopping a DMA transfer before it completes, this doesn't provide a means of detecting how many elements in `old_buf` are valid (in the case of MemoryToMemory or PeripheralToMemory). Perhaps we should also return the number of data actually transferred here? Alternative would be the number remaining, but they're essentially the same thing.

Reply:
Resolved in https://github.com/stm32-rs/stm32h7xx-hal/pull/173